### ISSUE TITLE
Documentation: Clarify rules re the 80 character limit in inline code comments

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#4666]
This came up in #4626, but there are multiple violations of the coding standard re line length and wrapping, especially in code we've inherited from Views 7.x. 

The specific rule is listed in one of the bullet points in the "General guidelines" section in  https://api.backdropcms.org/doc-standards#general, and it is currently phrased as follows:

> - Lines containing comments (including docblocks) must wrap as close to 80 characters as possible without going over, with a few exceptions (noted in the Tag Reference below).

Some people interpret this as "not longer than 80 characters - but shorter is not forbidden.", while others as "fit as many words as possible in a single line, but do not exceed 80 characters".

We should clarify this, and possibly add a good vs. bad examples in the documentation.

**Comment by klonos** [comment]
...with #3213 we need to get this sorted, so that we can have proper PHPCS rulesets in place.

**Comment by indigoxela** [comment]
> ... so that we can have proper PHPCS rulesets in place.

Good point, phpcs does *not* warn if comments are shorter. - But such a check might be hard to implement anyway.

**Comment by klonos** [comment]
> But such a check might be hard to implement anyway.

Yes, that's true. I'm wondering if we can have warnings instead of hard fails after checking. We can then do a manual check.

**Comment by indigoxela** [comment]
I'm all for clarifying that number. It could help to avoid exhausting discussions in PR comments and the issue queue.

**Comment by BWPanda** [comment]
@klonos To better understand the issue here, can you give examples in core where this standard is interpreted differently...?

**Comment by klonos** [comment]
@BWPanda there's dozens of examples, but usually the Views and Views Ui modules are where this happens the most. Here's an example (which also has double spaces after periods - a thing that [old-timers used to be taught in typing lessons **in typewriters**](https://www.cultofpedagogy.com/two-spaces-after-period) 😅 ): https://github.com/backdrop/backdrop/blob/1.x/core/modules/views/views.module#L423

```php
      // The menu system doesn't support having two otherwise
      // identical paths with different placeholders.  So we
      // want to remove the existing items from the menu whose
      // paths would conflict with ours.

      // First, we must find any existing menu items that may
      // conflict.  We use a regular expression because we don't
      // know what placeholders they might use.  Note that we
      // first construct the regex itself by replacing %views_arg
      // in the display path, then we use this constructed regex
      // (which will be something like '#^(foo/%[^/]*/bar)$#') to
      // search through the existing paths.
```

...should instead be:

```php
      // The menu system doesn't support having two otherwise identical paths
      // with different placeholders. So we want to remove the existing items
      // from the menu whose paths would conflict with ours.

      // First, we must find any existing menu items that may conflict. We use a
      // regular expression because we don't know what placeholders they might
      // use. Note that we first construct the regex itself by replacing
      // %views_arg in the display path, then we use this constructed regex
      // (which will be something like '#^(foo/%[^/]*/bar)$#') to search through
      // the existing paths.
```

**Comment by klonos** [comment]
PS: don't get me started on punctuation either on the last example 😅 😓 ...so when one reads this:

`This will automatically be called in order to 'load' a views argument`

...are they to interpret it as:

`This will automatically be called in order, to 'load' a views argument`

...or as:

`This will automatically be called, in order to 'load' a views argument`

(that's another issue though)

**Comment by BWPanda** [comment]
Hmm, ok. Seems as if these docblocks were using a different (shorter) line length back then...

**Comment by BWPanda** [comment]
I'm not sure this is a matter of interpreting the standards differently, but rather having different (or no?) standards when this was originally written (and then never brought up-to-standard).

**Comment by klonos** [comment]
> ...having different (or no?) standards when this was originally written (and then never brought up-to-standard).

Yup, this seems to be the case for the first example I've provided. The second seems to be intentional though ("for better readability" ?? 🤷 ). As I said, I've been seeing dozens and dozens of these over the years, and that's why I've been to fix any "surrounding" inline comments in the PRs I file (risking the wrath of code reviewers, telling me that I'm making scope-creeping changes - sorry).

**Comment by indigoxela** [comment]
Before this all becomes too dogmatic...

The *intention* of the text length limit is **readability**. Super long text lines are hard to read. Slightly shorter ones are fine to read.

Feel free to ignore this, it's just an opinion. The dogma wins.

**Comment by BWPanda** [comment]
@indigoxela Are you referring to the end of a sentence and then starting a new line before the 80 character mark:

```
This explains the text in the function below.
Another line of text starts on a new line, before 80 characters was reached above.
```

Or wrapping text in the middle of a sentence at a different character limit?

```
This text wraps
to fifteen
characters.
```

Extreme examples, but just trying to understand which situation you're promoting...

**Comment by indigoxela** [comment]
I'm referring to:
```
   *   Whether the active plugin does handle translation or not.
   *   FALSE for the plugin "none" or if no plugin was available.
```
or
```
        // If it is still NULL not even plugin "none" is available.
        // Only needed for edge cases like core update.
```
Both were OK for phpcs but have been rejected by a reviewer.
Which was a pretty frustrating situation for me (probably for both people involved).
Even more frustration came up when a line was 79 characters long and still was rejected.

If there's a _law_ that lines *may not be any shorter than 80 chars*, please document it somewhere.
If lines are allowed to be shorter than 80 chars, please document, too.
I just don't ever want to get involved in another pointless discussion regarding the 80-chars-dogma when providing a PR.

**Comment by BWPanda** [comment]
Ok, thanks @indigoxela. So yeah, seems we need to decide whether:

1. 80-character mark needs to be reached before starting a sentence on a new line, or
2. Sentences can start on a new line before the previous line reaches 80-characters.

And then clarify that decision in the documentation standards.

**Comment by docwilmot** [comment]
Can I propose in the meantime we agree NOT to reject otherwise valid code in PRs that are not specifically for documentation, and instead raise a separate issue? That would allow @indigoxela and others to get on with fixing things while we decide about this code standard, and we can do a META/bulk code standards fix in the future sometime.

**Comment by stpaultim** [comment]
If this has not come up in a DEV meeting, I recommend that we first bring it up there. Because, if we can come up with a good solution in a DEV meeting, it will be much less work (and it will happen much faster) than taking this to the PMC. Keeping in mind, that the solution in the DEV meeting might be, to escalate to the PMC for an official decision. 

I would like to talk about this in a DEV meeting, because in that situation I can get a better sense of how issues like this have been handled in the past (if there have been any). If this has not come up yet (yesterday), I'll try to remember to bring this up next week.

**Comment by quicksketch** [comment]
> Ok, thanks @indigoxela. So yeah, seems we need to decide whether:
> 
> 1. 80-character mark needs to be reached before starting a sentence on a new line, or
> 2. Sentences can start on a new line before the previous line reaches 80-characters.

I think officially allowing shorter line lengths for the purpose of readability (option 2) should be made coding policy.

**Comment by BWPanda** [comment]
Assuming there are no objections to @quicksketch's proposal, here are my suggested changes:

https://api.backdropcms.org/doc-standards#general

Change:
> Lines containing comments (including docblocks) must wrap as close to 80 characters as possible without going over, with a few exceptions (noted in the Tag Reference below).

to:
> Lines containing comments (including docblocks) must not exceed 80 characters, with a few exceptions (noted in the Tag Reference below). For readability purposes, lines *may* be shorter than 80 characters if necessary.

https://api.backdropcms.org/doc-standards#functions

Change:
> In the first line, declare what type of callback it is (Page callback, Access callback, etc.), and give a short description of what it does (like you would for any function). Total line length: 80 characters, ending in "."

to:
> In the first line, declare what type of callback it is (Page callback, Access callback, etc.), and give a short description of what it does (like you would for any function). Total line length: 80 characters or less, ending in "."

**Comment by BWPanda** [comment]
Technically there have been no objections, so I went ahead and made my suggested changes to https://docs.backdropcms.org/doc-standards#general
